neuralforecast/models/nhits/nhits.py

```diff
@@ -787,3 +787,137 @@ def forecast(self: NHITS, Y_df: pd.DataFrame, X_df: pd.DataFrame = None, S_df: p
+        gpus = -1 if t.cuda.is_available() else 0
+        trainer = pl.Trainer(progress_bar_refresh_rate=1,
+                             gpus=gpus,
+                             logger=False)
+
+    # Forecast all rolled windows
+    outputs = trainer.predict(self, loader)
+
+    y_true, y_hat, _ = [t.cat(output).cpu().numpy() for output in zip(*outputs)]
+
+    #------------------------------------ ds Wrangling -------------------------------------#
+
+    # Y_hat dataframe wrangling
+    unique_ids = Y_df['unique_id'].unique()
+
+    # Extend ds vector to match padded ts_tensor
+    timedelta = dataset.ds[1] -dataset.ds[0]
+    start_date = dataset.ds[0] - dataset.input_size*timedelta
+    end_date = dataset.ds[-1] + dataset.output_size*timedelta
+    ds = pd.date_range(start_date, end_date, freq=dataset.frequency)
+
+    # Padded sample_mask to filter forecasted datestamps
+    sample_mask = dataset.ts_tensor[:,dataset.t_cols.index('sample_mask'),:]
+    padding = (self.n_time_in, self.n_time_out)
+    padder = t.nn.ConstantPad1d(padding=padding, value=0)
+    sample_mask = padder(sample_mask)
+
+    # Create tensor with rolled date indexes and sample_mask
+    # [n_series,n_channel,n_time_padded]->[n_series,n_channel,n_windows,L+H]
+    ds_idxs = t.tile(t.arange(start=0, end=len(ds)), dims=[len(sample_mask), 1])
+    windows_tensor = t.cat((sample_mask[:,None,:], ds_idxs[:,None,:]), dim=1)
+    windows_tensor = windows_tensor.unfold(dimension=-1,
+                            size=self.n_time_in+self.n_time_out,
+                            step=dataset.sample_freq) # For the moment sample_freq=1
+
+    # Sampling condition, ALL ds in the forecast window should be sampleaple
+    # [n_series,n_windows,L+H]->[n_series,n_windows]
+    sample_mask_rolled = windows_tensor[:,0,:,-self.n_time_out:]
+    sample_condition = (sample_mask_rolled.sum(dim=2)==self.n_time_out)
+
+    # Use date indexes to create windows of date stamps
+    # And filter the dates of the forecast and the forecast creation date
+    # with -self.n_time_out-1
+    ds_rolled = ds[windows_tensor[:,1,:,-self.n_time_out-1:].numpy().astype(int)]
+
+    # Filter rolled ds with sample_condition
+    ds_rolled = ds_rolled.reshape(-1, self.n_time_out+1)
+    ds_rolled = ds_rolled[sample_condition.flatten()]
+
+    # Split forecast creation date from the prediction ds
+    fcds = ds_rolled[:,0]
+    ds_rolled = ds_rolled[:,1:]
+
+    #------------------------------------- Forecast df -------------------------------------#
+
+    # Compute per series' observations = horizon * n_windows
+    # and create unique_id column in long format
+    unique_id_counts = self.n_time_out*sample_condition.sum(dim=1) # vectorized
+    unique_ids_column = [[u_id]*count for u_id,count in zip(unique_ids, unique_id_counts)]
+    unique_ids_column = np.concatenate(unique_ids_column)
+
+    # Forecast creation date column in long format, shared across horizon
+    fcds_column = fcds.repeat(self.n_time_out)
```

Comment:
These lines should be a method of the class `TimeSeriesDataset` called `get_fcds_columns` or something similar.  Particularly to avoid duplicating code in each model.